waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,311 @@
+
+      #TODO send error code and desc to client
+      return
+
+    if hashes.len > 0 and self.transferCallBack.isSome():
+      let callback = self.transferCallBack.get()
+
+      (await callback(hashes, conn.peerId)).isOkOr:
+        debug "transfer callback failed", error
+
+    debug "sync session ended gracefully",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId
+
+  self.handler = handle
+  self.codec = WakuSyncCodec
+
+proc new*(
+    T: type WakuSync,
+    peerManager: PeerManager,
+    maxFrameSize: int = DefaultMaxFrameSize,
+    syncInterval: timer.Duration = DefaultSyncInterval,
+    relayJitter: Duration = DefaultGossipSubJitter,
+    pruneCB: Option[PruneCallBack] = none(PruneCallback),
+    transferCB: Option[TransferCallback] = none(TransferCallback),
+): Result[T, string] =
+  let storage = Storage.new().valueOr:
+    return err("negentropy storage creation failed")
+
+  let sync = WakuSync(
+    storage: storage,
+    peerManager: peerManager,
+    maxFrameSize: maxFrameSize,
+    syncInterval: syncInterval,
+    relayJitter: relayJitter,
+    pruneCallBack: pruneCB,
+    pruneOffset: syncInterval div 2,
+    transferCallBack: transferCB,
+  )
+
+  sync.initProtocolHandler()
+
+  info "WakuSync protocol initialized"
+
+  return ok(sync)
+
+proc periodicSync(self: WakuSync, callback: TransferCallback) {.async.} =
+  debug "periodic sync initialized", interval = $self.syncInterval
+
+  while true:
+    await sleepAsync(self.syncInterval)
+
+    debug "periodic sync started"
+
+    let (hashes, peer) = (await self.sync()).valueOr:
+      debug "sync failed", error
+      continue
+
+    if hashes.len > 0:
+      (await callback(hashes, peer.peerId)).isOkOr:
+        debug "transfer callback failed", error
+        continue
+
+    debug "periodic sync done", hashSynced = hashes.len
+
+proc periodicPrune(self: WakuSync, callback: PruneCallback) {.async.} =
+  debug "periodic prune initialized", interval = $self.syncInterval
+
+  await sleepAsync(self.syncInterval)
+
+  var pruneStop = getNowInNanosecondTime()
+
+  while true:
+    await sleepAsync(self.syncInterval)
+
+    debug "periodic prune started",
+      startTime = self.pruneStart - self.pruneOffset.nanos,
+      endTime = pruneStop,
+      storageSize = self.storage.len
+
+    var (elements, cursor) =
+      (newSeq[(WakuMessageHash, Timestamp)](0), none(WakuMessageHash))
+
+    var tries = 3
+    while true:
+      (elements, cursor) = (
+        await callback(self.pruneStart - self.pruneOffset.nanos, pruneStop, cursor)
+      ).valueOr:
+        debug "pruning callback failed", error
+        if tries > 0:
+          tries -= 1
+          await sleepAsync(30.seconds)
+          continue
+        else:
+          break
+
+      if elements.len == 0:
+        break
+
+      for (hash, timestamp) in elements:
+        self.storage.erase(timestamp, hash).isOkOr:
```

Comment:
Cool, I found where we delete data :D
I think it could be interesting to implement a tiny unit-test where we ingest many messages, then delete, and then validate that the occupied memory didn't grow.

Reply:
Ah yes good idea!

Reply:
:thinking: any ideas how to check memory size?

Reply:
> 🤔 any ideas how to check memory size?

We can try with `getOccupiedMem()`, from `system` module, which gives the current memory in bytes.
We can measure at three points:
1. At the beginning
2. After allocating memory 
3. At the end, to make sure the memory is back to the level at 1.

Reply:
`getOccupiedMem()` seams to count only what is managed by the GC. Since we use Negentropy C++ lib that memory is managed on the C++ side. Any other ideas @Ivansete-status ? IDK what to do...

Reply:
What do we target here with mem size check?
If this lib is optimized it might not shrink memory that way. Also it probably preallocates to some extent as it is pretty costly to grow in small chunks also to shrink. Maybe we need to review the library itself?